pkg/kvstore/etcd.go

```diff
@@ -0,0 +1,372 @@
+	return err
+}
+
+func (e *EtcdClient) InitializeFreeID(path string, firstID uint32) error {
+	kvLocker, err := e.LockPath(path)
+	if err != nil {
+		return err
+	}
+	defer kvLocker.Unlock()
+
+	log.Debug("Trying to acquire free ID...")
+	k, err := e.GetValue(path)
+	if err != nil {
+		return err
+	}
+	if k != nil {
+		// FreeID already set
+		return nil
+	}
+	log.Info("Trying to put free ID...")
+	err = e.SetValue(path, firstID)
+	if err != nil {
+		return err
+	}
+	log.Info("Free ID for path %s successfully initialized", path)
+
+	return nil
+}
+
+func (e *EtcdClient) GetMaxID(key string, firstID uint32) (uint32, error) {
+	value, err := e.GetValue(key)
+	if err != nil {
+		return 0, err
+	}
+	if value == nil {
+		// FreeID is empty? We should set it out!
+		log.Infof("Empty FreeID, setting it up with default value %d", firstID)
+		if err := e.InitializeFreeID(key, firstID); err != nil {
+			return 0, err
+		}
+		value, err = e.GetValue(key)
+		if err != nil {
+			return 0, err
+		}
+		if value == nil {
+			// Something is really wrong
+			errMsg := "Unable to retrieve last free ID because the key is always empty\n"
+			log.Errorf(errMsg)
+			return 0, fmt.Errorf(errMsg)
+		}
+	}
+	var freeID uint32
+	if err := json.Unmarshal(value, &freeID); err != nil {
+		return 0, err
+	}
+	log.Debugf("Retrieving max free ID %d", freeID)
+	return freeID, nil
+}
+
+func (e *EtcdClient) SetMaxID(key string, firstID, maxID uint32) error {
+	value, err := e.GetValue(key)
+	if err != nil {
+		return err
+	}
+	if value == nil {
+		// FreeID is empty? We should set it out!
+		log.Infof("Empty FreeID, setting it up with default value %d", firstID)
+		if err := e.InitializeFreeID(key, firstID); err != nil {
+			return err
+		}
+		k, err := e.GetValue(key)
+		if err != nil {
+			return err
+		}
+		if k == nil {
+			// Something is really wrong
+			errMsg := "Unable to setting ID because the key is always empty\n"
+			log.Errorf(errMsg)
+			return fmt.Errorf(errMsg)
+		}
+	}
+	return e.SetValue(key, maxID)
+}
+
+func (e *EtcdClient) GASNewSecLabelID(basePath string, baseID uint32, secCtxLabels *types.SecCtxLabel) error {
+	setID2Label := func(lockPair KVLocker) error {
+		defer lockPair.Unlock()
+		secCtxLabels.ID = baseID
+		keyPath := path.Join(basePath, strconv.FormatUint(uint64(secCtxLabels.ID), 10))
+		if err := e.SetValue(keyPath, secCtxLabels); err != nil {
+			return err
+		}
+		return e.setMaxLabelID(baseID + 1)
+	}
+
+	beginning := baseID
+	for {
+		log.Debugf("Trying to acquire a new free ID %d", baseID)
+		keyPath := path.Join(basePath, strconv.FormatUint(uint64(baseID), 10))
+
+		locker, err := e.LockPath(GetLockPath(keyPath))
+		if err != nil {
+			return err
+		}
+
+		value, err := e.GetValue(keyPath)
+		if err != nil {
+			locker.Unlock()
+			return err
+		}
+		if value == nil {
+			return setID2Label(locker)
+		}
+		var consulLabels types.SecCtxLabel
+		if err := json.Unmarshal(value, &consulLabels); err != nil {
+			locker.Unlock()
```

Comment:
Seems quite error prone to unlock this manually on so many early returns. This looks like a good candidate to extract to a function(s) in order to be able to use defer.